core/src/main/java/com/linecorp/armeria/server/auth/AuthServiceBuilder.java

```diff
@@ -43,8 +40,24 @@ public final class AuthServiceBuilder {
         if (cause != null) {
             AuthService.logger.warn("Unexpected exception during authorization.", cause);
         }
+        if (this.authType.equals(AuthType.OAUTH2)) {
```

Comment:
When several `Authorizer`s can be added to an `AuthService`, `authType` can't handle the case.
What do you think of checking the `Authorization` header to know if the request uses basic or oauth2 authorization?

```java
String authorization =  req.headers().get(HttpHeaderNames.AUTHORIZATION, "");
if (!authorization.isEmpty()) {
  if (BasicTokenExtractor.AUTHORIZATION_HEADER_PATTERN.matches(authorization)) {
      // return a response with `WWW_AUTHENTICATE: Basic`
  }

  if (OAuth2TokenExtractor.AUTHORIZATION_HEADER_PATTERN.matches(authorization)) {
      // return a response with `WWW_AUTHENTICATE: Bearer`
  }
}

return HttpResponse.of(HttpStatus.UNAUTHORIZED);
```

Reply:
I prefer setting `AuthType` to `ctx.attr()` in `DelegatingHttpRequestAuthorizer` rather than fixing `AuthServiceBuilder`. For example:
```java
final class DelegatingHttpRequestAuthorizer<T> extends AbstractAuthorizerWithHandlers<HttpRequest> {

    private final Function<? super RequestHeaders, @Nullable T> tokenExtractor;
    private final Authorizer<? super T> delegate;
    private final AuthType authType;

    DelegatingHttpRequestAuthorizer(Function<? super RequestHeaders, @Nullable T> tokenExtractor,
                                    Authorizer<? super T> delegate) {
        this.tokenExtractor = requireNonNull(tokenExtractor, "tokenExtractor");
        this.delegate = requireNonNull(delegate, "authorizer");
        if (tokenExtractor instanceof BasicTokenExtractor) {
            authType = AuthType.BASIC;
        } else if (...) { ...}
    }

    @Override
    public CompletionStage<AuthorizationStatus> authorizeAndSupplyHandlers(ServiceRequestContext ctx,
                                                                           @Nullable HttpRequest request) {
        final T token = tokenExtractor.apply(requireNonNull(request, "request").headers());
        ctx.setAttr(AUTH_TYPE_KEY, authType);
        return delegate.authorizeAndSupplyHandlers(ctx, token);
    }
}
```
If `AuthType` is set to `ctx.attrs()`, we can use the value to know the last authorizer applied.
```java
// in AuthServiceBulder
private AuthFailureHandler failureHandler = (delegate, ctx, req, cause) -> {
    if (cause != null) {
        AuthService.logger.warn("Unexpected exception during authorization.", cause);
    }
    AuthType authType = ctx.attr(AUTH_TYPE_KEY);
    if (authType = AuthType.BASIC) {
        // return a response with `WWW_AUTHENTICATE: Basic`
    }
    ...
}
```